session.go

```diff
@@ -0,0 +1,526 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package iotdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"github.com/apache/iotdb-client-go/rpc"
+	"github.com/apache/thrift/lib/go/thrift"
+	"net"
+	"time"
+)
+
+const protocolVersion = rpc.TSProtocolVersion_IOTDB_SERVICE_PROTOCOL_V3
+
+var lengthError = errors.New("this is a new error")
+
+func (s *Session) Open(enableRPCCompression bool, connectionTimeoutInMs int) error {
+	var protocolFactory thrift.TProtocolFactory
+	var err error
+	s.trans, err = thrift.NewTSocketTimeout(net.JoinHostPort(s.config.Host, s.config.Port), time.Duration(connectionTimeoutInMs))
+	if err != nil {
+		return err
+	}
+	s.trans = thrift.NewTFramedTransport(s.trans)
+	if !s.trans.IsOpen() {
+		err = s.trans.Open()
+		if err != nil {
+			return err
+		}
+	}
+	if enableRPCCompression {
+		protocolFactory = thrift.NewTCompactProtocolFactory()
+	} else {
+		protocolFactory = thrift.NewTBinaryProtocolFactoryDefault()
+	}
+	iProtocol := protocolFactory.GetProtocol(s.trans)
+	oProtocol := protocolFactory.GetProtocol(s.trans)
+	s.client = rpc.NewTSIServiceClient(thrift.NewTStandardClient(iProtocol, oProtocol))
+	s.config.ZoneId = DefaultZoneId
+	tSOpenSessionReq := rpc.TSOpenSessionReq{ClientProtocol: protocolVersion, ZoneId: s.config.ZoneId,
+		Username: &s.config.User, Password: &s.config.Passwd}
+	tSOpenSessionResp, err := s.client.OpenSession(context.Background(), &tSOpenSessionReq)
+	if err != nil {
+		return err
+	}
+	s.sessionId = tSOpenSessionResp.GetSessionId()
+	s.requestStatementId, err = s.client.RequestStatementId(context.Background(), s.sessionId)
+	s.config.FetchSize = DefaultFetchSize
+	if err != nil {
+		return err
+	}
+	if s.config.ZoneId != "" {
+		s.SetTimeZone(s.config.ZoneId)
+	} else {
+		s.config.ZoneId, err = s.GetTimeZone()
+	}
+	return err
+}
+
+func (s *Session) CheckTimeseriesExists(path string) bool {
+	dataSet, _ := s.ExecuteQueryStatement("SHOW TIMESERIES " + path)
+	result := dataSet.HasNext()
+	dataSet.CloseOperationHandle()
+	return result
+}
+
+func (s *Session) Close() error {
+	tSCloseSessionReq := rpc.NewTSCloseSessionReq()
+	tSCloseSessionReq.SessionId = s.sessionId
+	_, err := s.client.CloseSession(context.Background(), tSCloseSessionReq)
+	s.trans.Close()
+	return err
+}
+
+/*
+ *set one storage group
+ *
+ *param
+ *storageGroupId: string, storage group name (starts from root)
+ *
+ */
+func (s *Session) SetStorageGroup(storageGroupId string) error {
+	_, err := s.client.SetStorageGroup(context.Background(), s.sessionId, storageGroupId)
```

Comment:
need to check `status.code ! = 200`, it is noticeable that the server returns an error when it repeatedly executes the `SetStorageGroup` method.

Reply:
Thank you for your advice